cmd/forwarder/http.go

```diff
@@ -87,76 +95,83 @@ func extractRemoteAddr(r *http.Request) string {
 	return remoteAddr
 }
 
-func (s *httpServer) Run() error {
-	go s.awaitShutdown()
+func (s *httpServer) handleLogs(w http.ResponseWriter, r *http.Request) {
+	defer s.posts.UpdateSince(time.Now())
 
-	//FXME: check outlet depth?
-	http.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
-		defer s.healthChecks.UpdateSince(time.Now())
-		if s.isShuttingDown {
-			http.Error(w, "Shutting down", 503)
-			return
-		}
+	if s.Config.EnforceSsl && r.Header.Get("X-Forwarded-Proto") != "https" {
+		s.handleHTTPError(w, "Only SSL requests accepted", 400)
+		return
+	}
 
-	})
+	if s.isShuttingDown {
+		s.handleHTTPError(w, "Shutting down", 503)
+		return
+	}
 
-	http.HandleFunc("/logs", func(w http.ResponseWriter, r *http.Request) {
-		defer s.posts.UpdateSince(time.Now())
+	if r.Method != "POST" {
+		s.handleHTTPError(w, "Only POST is accepted", 400)
+		return
+	}
 
-		if s.Config.EnforceSsl && r.Header.Get("X-Forwarded-Proto") != "https" {
-			s.handleHTTPError(w, "Only SSL requests accepted", 400)
-			return
-		}
+	contentType := r.Header.Get("Content-Type")
+	if fixers[contentType] == nil {
+		s.handleHTTPError(w, "Unsupported Content-Type", 400)
+		return
+	}
 
-		if s.isShuttingDown {
-			s.handleHTTPError(w, "Shutting down", 503)
-			return
-		}
+	if !s.auth.Authenticate(r) {
+		s.pAuthErrors.Inc(1)
+		s.handleHTTPError(w, "Unable to authenticate request", 401)
+		return
+	}
 
-		if r.Method != "POST" {
-			s.handleHTTPError(w, "Only POST is accepted", 400)
-			return
-		}
+	s.pAuthSuccesses.Inc(1)
 
-		if r.Header.Get("Content-Type") != "application/logplex-1" {
-			s.handleHTTPError(w, "Only Content-Type application/logplex-1 is accepted", 400)
-			return
-		}
+	remoteAddr := extractRemoteAddr(r)
+	requestID := r.Header.Get("X-Request-Id")
+	drainToken := r.Header.Get("Logplex-Drain-Token")
+
+	body := r.Body
+	defer body.Close()
+	var err error
 
-		if !s.auth.Authenticate(r) {
-			s.pAuthErrors.Inc(1)
-			s.handleHTTPError(w, "Unable to authenticate request", 401)
+	if r.Header.Get("Content-Encoding") == "gzip" {
+		body, err = gzip.NewReader(r.Body)
+		if err != nil {
+			s.handleHTTPError(w, "Could not decode gzip request", 500)
 			return
-		} else {
-			s.pAuthSuccesses.Inc(1)
 		}
+	}
 
-		remoteAddr := extractRemoteAddr(r)
-		requestID := r.Header.Get("X-Request-Id")
-		logplexDrainToken := r.Header.Get("Logplex-Drain-Token")
+	var fixedBody []byte
+	fixedBody, err = fixers[contentType](body, remoteAddr, drainToken)
```

Comment:
Should actually running the fixer stay in the process method? Moving it here seems to lose some logging context.

Reply:
Sure! I don't feel strongly about where its called from. I think I prefer the way I have it here (the concrete A then B calls/steps over nesting step A inside step B). I'll experiment with putting it inside process and see how it feels.